src/probe/probe.c

```diff
@@ -1099,23 +1152,69 @@ int probe_point_add(uint32_t count, struct probe_point *probe)
 		_probe->probe_points[first_free].stream_tag =
 			probe[i].stream_tag;
 
-		if (probe[i].purpose == PROBE_PURPOSE_LOGGING) {
+		if (probe[i].purpose == PROBE_PURPOSE_LOGGING || probe[i].buffer_id.fields.module_id == 0) {
 #if CONFIG_LOG_BACKEND_SOF_PROBE
 			probe_logging_init(probe_logging_hook);
 #else
 			return -EINVAL;
 #endif
 		} else {
-			notifier_register(_probe, dev->cb, NOTIFIER_ID_BUFFER_PRODUCE,
+#if CONFIG_IPC_MAJOR_4
+			notifier_register(&probe[i].buffer_id.full_id, buf, NOTIFIER_ID_BUFFER_PRODUCE,
+					  &probe_cb_produce, 0);
+			notifier_register(&probe[i].buffer_id.full_id, buf, NOTIFIER_ID_BUFFER_FREE,
+					  &probe_cb_free, 0);
+#else
+			notifier_register(&probe[i].buffer_id.full_id, dev->cb, NOTIFIER_ID_BUFFER_PRODUCE,
 					  &probe_cb_produce, 0);
-			notifier_register(_probe, dev->cb, NOTIFIER_ID_BUFFER_FREE,
+			notifier_register(&probe[i].buffer_id.full_id, dev->cb, NOTIFIER_ID_BUFFER_FREE,
 					  &probe_cb_free, 0);
+#endif
 		}
 	}
 
 	return 0;
 }
 
+#if CONFIG_IPC_MAJOR_3
+int probe_dma_info(struct sof_ipc_probe_info_params *data, uint32_t max_size)
+{
+	struct probe_pdata *_probe = probe_get();
+	uint32_t i = 0;
+	uint32_t j = 0;
+
+	tr_dbg(&pr_tr, "probe_dma_info()");
+
+	if (!_probe) {
+		tr_err(&pr_tr, "probe_dma_info(): Not initialized.");
+
+		return -EINVAL;
+	}
+
+	data->rhdr.hdr.size = sizeof(*data);
+
+	/* search all injection DMAs to send them in reply */
+	while (i < CONFIG_PROBE_DMA_MAX &&
+	       data->rhdr.hdr.size + sizeof(struct probe_dma) < max_size) {
```

Comment:
a `for` loop seems to fit better here

Reply:
old ipc3 code, just moved to another place